File: analyze_challenge_cards.py
#!/usr/bin/env python3
"""
Analyze all 240 challenge cards from lessons_all_pillars.sql
Check for:
- Specificity (concrete actions vs vague goals)
- Time-bound nature (<10 minutes)
- Connection to lesson concept
- Clear success criteria
"""

import re
import json

def extract_lessons_from_sql(filepath):
    """Extract all lesson records from SQL file."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Find all INSERT INTO lessons statements
    # Pattern: INSERT INTO lessons ... VALUES (lesson_data), (lesson_data), ...
    lessons = []
    
    # Split by INSERT INTO lessons statements
    inserts = re.findall(
        r"INSERT INTO lessons \([^)]+\) VALUES\s*\n(.*?)(?=INSERT INTO|-- ===|$)",
        content,
        re.DOTALL
    )
    
    for insert_block in inserts:
        # Split individual lesson records - each starts with ('33333333...
        lesson_matches = re.findall(
            r"\('33333333-[^']+','[^']+','([^']+)',(\d+),\s*'((?:[^']|'')*)',\s*'((?:[^']|'')*)',\s*'((?:[^']|'')*)',\s*'((?:[^']|'')*)',\s*'((?:[^']|'')*?)'\)",
            insert_block,
            re.DOTALL
        )
        
        for match in lesson_matches:
            title, display_order, concept, example, mistake, science, challenge = match
            lessons.append({
                'title': title,
                'display_order': int(display_order),
                'card_challenge': challenge.replace("''", "'")  # Unescape SQL quotes
            })
    
    return lessons

def analyze_challenge(challenge_text, title):
    """Analyze a single challenge card."""
    issues = []
    
    # Check 1: Length (should be 50-75 words per design doc)
    word_count = len(challenge_text.split())
    if word_count < 30:
        issues.append(f"Too short ({word_count} words) - may lack detail")
    elif word_count > 100:
        issues.append(f"Too long ({word_count} words) - may be overwhelming")
    
    # Check 2: Immediacy triggers (right now, today, next time, etc.)
    immediacy_triggers = ['right now', 'today', 'next time', 'this week', 'before you', 'in your next', 'set a timer']
    has_trigger = any(trigger in challenge_text.lower() for trigger in immediacy_triggers)
    if not has_trigger:
        issues.append("Missing immediacy trigger (right now, today, next time)")
    
    # Check 3: Time-bound language
    time_indicators = ['minute', 'second', 'hour', 'day', 'week', 'tomorrow', 'tonight']
    has_time = any(indicator in challenge_text.lower() for indicator in time_indicators)
    if not has_time:
        issues.append("Not clearly time-bound")
    
    # Check 4: Specific action verbs
    action_verbs = ['write', 'record', 'call', 'send', 'ask', 'do', 'try', 'practice', 'track', 'list', 'set', 'count', 'notice', 'say', 'tell', 'read', 'open']
    has_action = any(verb in challenge_text.lower() for verb in action_verbs)
    if not has_action:
        issues.append("Lacks clear action verb")
    
    # Check 5: Reflection prompt ("notice", "observe", "pay attention")
    reflection_prompts = ['notice', 'observe', 'pay attention', 'watch', 'feel', 'see what happens']
    has_reflection = any(prompt in challenge_text.lower() for prompt in reflection_prompts)
    if not has_reflection:
        issues.append("Missing reflection prompt")
    
    # Check 6: Vague language to avoid
    vague_words = ['try to', 'work on', 'think about', 'consider', 'start thinking', 'be more']
    has_vague = any(vague in challenge_text.lower() for vague in vague_words)
    if has_vague:
        issues.append("Contains vague language")
    
    # Check 7: Success metrics (numbers, completion indicators)
    has_metric = bool(re.search(r'\d+|count|number|baseline|measure|how many', challenge_text.lower()))
    
    return {
        'title': title,
        'word_count': word_count,
        'has_immediacy': has_trigger,
        'has_time_bound': has_time,
        'has_action_verb': has_action,
        'has_reflection': has_reflection,
        'has_vague_language': has_vague,
        'has_success_metric': has_metric,
        'issues': issues,
        'challenge_text': challenge_text[:200] + '...' if len(challenge_text) > 200 else challenge_text
    }

def generate_report(lessons):
    """Generate analysis report."""
    print("=" * 80)
    print("CHALLENGE CARD ACTIONABILITY REVIEW - 240 LESSONS")
    print("=" * 80)
    print()
    
    all_issues = []
    total_lessons = len(lessons)
    
    # Aggregate statistics
    stats = {
        'has_immediacy': 0,
        'has_time_bound': 0,
        'has_action_verb': 0,
        'has_reflection': 0,
        'has_vague_language': 0,
        'has_success_metric': 0,
        'word_count_ok': 0,
        'no_issues': 0
    }
    
    for lesson in lessons:
        analysis = analyze_challenge(lesson['card_challenge'], lesson['title'])
        
        if analysis['has_immediacy']: stats['has_immediacy'] += 1
        if analysis['has_time_bound']: stats['has_time_bound'] += 1
        if analysis['has_action_verb']: stats['has_action_verb'] += 1
        if analysis['has_reflection']: stats['has_reflection'] += 1
        if analysis['has_vague_language']: stats['has_vague_language'] += 1
        if analysis['has_success_metric']: stats['has_success_metric'] += 1
        if 30 <= analysis['word_count'] <= 100: stats['word_count_ok'] += 1
        if len(analysis['issues']) == 0: stats['no_issues'] += 1
        
        if analysis['issues']:
            all_issues.append(analysis)
    
    # Print summary statistics
    print(f"SUMMARY STATISTICS ({total_lessons} lessons analyzed)")
    print("-" * 80)
    print(f"✓ Immediacy trigger present:    {stats['has_immediacy']:3d} / {total_lessons} ({stats['has_immediacy']/total_lessons*100:.1f}%)")
    print(f"✓ Time-bound language:          {stats['has_time_bound']:3d} / {total_lessons} ({stats['has_time_bound']/total_lessons*100:.1f}%)")
    print(f"✓ Clear action verb:            {stats['has_action_verb']:3d} / {total_lessons} ({stats['has_action_verb']/total_lessons*100:.1f}%)")
    print(f"✓ Reflection prompt:            {stats['has_reflection']:3d} / {total_lessons} ({stats['has_reflection']/total_lessons*100:.1f}%)")
    print(f"✓ Success metric/measurement:   {stats['has_success_metric']:3d} / {total_lessons} ({stats['has_success_metric']/total_lessons*100:.1f}%)")
    print(f"✓ Appropriate word count:       {stats['word_count_ok']:3d} / {total_lessons} ({stats['word_count_ok']/total_lessons*100:.1f}%)")
    print(f"✗ Contains vague language:      {stats['has_vague_language']:3d} / {total_lessons} ({stats['has_vague_language']/total_lessons*100:.1f}%)")
    print(f"✓ NO ISSUES FOUND:              {stats['no_issues']:3d} / {total_lessons} ({stats['no_issues']/total_lessons*100:.1f}%)")
    print()
    
    # Print detailed issues
    if all_issues:
        print(f"CHALLENGES WITH ISSUES ({len(all_issues)} lessons need attention)")
        print("-" * 80)
        for i, issue_data in enumerate(all_issues[:20], 1):  # Show first 20
            print(f"\n{i}. {issue_data['title']}")
            print(f"   Word count: {issue_data['word_count']}")
            for issue in issue_data['issues']:
                print(f"   ⚠ {issue}")
            print(f"   Challenge: {issue_data['challenge_text']}")
        
        if len(all_issues) > 20:
            print(f"\n... and {len(all_issues) - 20} more challenges with issues")
    else:
        print("✓ ALL CHALLENGES PASS QUALITY CHECKS")
    
    print()
    print("=" * 80)
    print("RECOMMENDATIONS")
    print("=" * 80)
    print()
    
    if stats['has_immediacy'] < total_lessons * 0.9:
        print("• Add immediacy triggers (right now, today, next time) to more challenges")
    
    if stats['has_time_bound'] < total_lessons * 0.85:
        print("• Make challenges more time-bound (<10 minutes ideal)")
    
    if stats['has_reflection'] < total_lessons * 0.85:
        print("• Add reflection prompts (notice what happens, observe, pay attention)")
    
    if stats['has_vague_language'] > total_lessons * 0.1:
        print("• Replace vague language (try to, work on) with specific actions")
    
    if stats['has_success_metric'] < total_lessons * 0.5:
        print("• Add clearer success criteria so users know when they've completed the challenge")
    
    if stats['no_issues'] >= total_lessons * 0.8:
        print("✓ Challenges are in excellent shape - most meet all quality criteria")
    
    print()

def main():
    lessons = extract_lessons_from_sql('ascevo/supabase/lessons_all_pillars.sql')
    print(f"Extracted {len(lessons)} lessons from SQL file\n")
    generate_report(lessons)

if __name__ == '__main__':
    main()

File: analyze_challenge_cards.js
#!/usr/bin/env node
/**
 * Analyze all 240 challenge cards from lessons_all_pillars.sql
 * Check for:
 * - Specificity (concrete actions vs vague goals)
 * - Time-bound nature (<10 minutes)
 * - Connection to lesson concept
 * - Clear success criteria
 */

const fs = require('fs');

function extractLessonsFromSQL(filepath) {
  const content = fs.readFileSync(filepath, 'utf-8');
  
  // Match INSERT INTO lessons statements and extract challenge cards
  const lessons = [];
  
  // Split by lines and parse SQL INSERT statements
  const lines = content.split('\n');
  let currentLesson = null;
  let collectingChallenge = false;
  let challengeText = '';
  let parenDepth = 0;
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    
    // Check if this is a lesson INSERT line (starts with lesson data)
    if (line.match(/^\('33333333-/)) {
      // Parse the line to extract title and challenge
      // Format: ('id','unit_id','title',display_order,'concept','example','mistake','science','challenge')
      
      // Use a simple state machine to extract fields
      let inString = false;
      let field = '';
      let fields = [];
      let escapeNext = false;
      
      for (let j = 0; j < line.length; j++) {
        const char = line[j];
        
        if (escapeNext) {
          field += char;
          escapeNext = false;
          continue;
        }
        
        if (char === "'" && line[j + 1] === "'") {
          field += "'";
          escapeNext = true;
          continue;
        }
        
        if (char === "'") {
          inString = !inString;
          if (!inString && field) {
            fields.push(field);
            field = '';
          }
          continue;
        }
        
        if (inString) {
          field += char;
        }
      }
      
      if (fields.length >= 7) {
        const title = fields[2];
        const challenge = fields[6];
        
        currentLesson = { title, challenge };
        lessons.push(currentLesson);
      }
    }
  }
  
  return lessons;
}

function analyzeChallenge(challengeText, title) {
  const issues = [];
  
  // Check 1: Length (should be 50-75 words per design doc)
  const wordCount = challengeText.split(/\s+/).length;
  if (wordCount < 30) {
    issues.push(`Too short (${wordCount} words) - may lack detail`);
  } else if (wordCount > 100) {
    issues.push(`Too long (${wordCount} words) - may be overwhelming`);
  }
  
  // Check 2: Immediacy triggers
  const immediacyTriggers = ['right now', 'today', 'next time', 'this week', 'before you', 'in your next', 'set a timer', 'tonight', 'tomorrow'];
  const hasImmediacyTrigger = immediacyTriggers.some(trigger => 
    challengeText.toLowerCase().includes(trigger)
  );
  if (!hasImmediacyTrigger) {
    issues.push('Missing immediacy trigger (right now, today, next time)');
  }
  
  // Check 3: Time-bound language
  const timeIndicators = ['minute', 'second', 'hour', 'day', 'week', 'tomorrow', 'tonight'];
  const hasTimebound = timeIndicators.some(indicator => 
    challengeText.toLowerCase().includes(indicator)
  );
  if (!hasTimebound) {
    issues.push('Not clearly time-bound');
  }
  
  // Check 4: Specific action verbs
  const actionVerbs = ['write', 'record', 'call', 'send', 'ask', 'do', 'try', 'practice', 'track', 'list', 'set', 'count', 'notice', 'say', 'tell', 'read', 'open', 'identify', 'pick', 'answer', 'take', 'go'];
  const hasActionVerb = actionVerbs.some(verb => 
    challengeText.toLowerCase().includes(verb)
  );
  if (!hasActionVerb) {
    issues.push('Lacks clear action verb');
  }
  
  // Check 5: Reflection prompt
  const reflectionPrompts = ['notice', 'observe', 'pay attention', 'watch', 'feel', 'see what happens'];
  const hasReflection = reflectionPrompts.some(prompt => 
    challengeText.toLowerCase().includes(prompt)
  );
  if (!hasReflection) {
    issues.push('Missing reflection prompt (notice, observe)');
  }
  
  // Check 6: Vague language to avoid
  const vagueWords = ['try to', 'work on', 'think about', 'consider', 'start thinking', 'be more'];
  const hasVague = vagueWords.some(vague => 
    challengeText.toLowerCase().includes(vague)
  );
  if (hasVague) {
    issues.push('Contains vague language (try to, work on, think about)');
  }
  
  // Check 7: Success metrics
  const hasMetric = /\d+|count|number|baseline|measure|how many/i.test(challengeText);
  
  return {
    title,
    wordCount,
    hasImmediacy: hasImmediacyTrigger,
    hasTimebound,
    hasActionVerb,
    hasReflection,
    hasVague,
    hasMetric,
    issues,
    challengePreview: challengeText.length > 200 ? challengeText.substring(0, 200) + '...' : challengeText
  };
}

function generateReport(lessons) {
  console.log('='.repeat(80));
  console.log('CHALLENGE CARD ACTIONABILITY REVIEW - ALL LESSONS');
  console.log('='.repeat(80));
  console.log();
  
  const stats = {
    hasImmediacy: 0,
    hasTimebound: 0,
    hasActionVerb: 0,
    hasReflection: 0,
    hasVague: 0,
    hasMetric: 0,
    wordCountOk: 0,
    noIssues: 0
  };
  
  const allIssues = [];
  
  for (const lesson of lessons) {
    const analysis = analyzeChallenge(lesson.challenge, lesson.title);
    
    if (analysis.hasImmediacy) stats.hasImmediacy++;
    if (analysis.hasTimebound) stats.hasTimebound++;
    if (analysis.hasActionVerb) stats.hasActionVerb++;
    if (analysis.hasReflection) stats.hasReflection++;
    if (analysis.hasVague) stats.hasVague++;
    if (analysis.hasMetric) stats.hasMetric++;
    if (analysis.wordCount >= 30 && analysis.wordCount <= 100) stats.wordCountOk++;
    if (analysis.issues.length === 0) stats.noIssues++;
    
    if (analysis.issues.length > 0) {
      allIssues.push(analysis);
    }
  }
  
  const total = lessons.length;
  
  console.log(`SUMMARY STATISTICS (${total} lessons analyzed)`);
  console.log('-'.repeat(80));
  console.log(`✓ Immediacy trigger present:    ${stats.hasImmediacy.toString().padStart(3)} / ${total} (${(stats.hasImmediacy/total*100).toFixed(1)}%)`);
  console.log(`✓ Time-bound language:          ${stats.hasTimebound.toString().padStart(3)} / ${total} (${(stats.hasTimebound/total*100).toFixed(1)}%)`);
  console.log(`✓ Clear action verb:            ${stats.hasActionVerb.toString().padStart(3)} / ${total} (${(stats.hasActionVerb/total*100).toFixed(1)}%)`);
  console.log(`✓ Reflection prompt:            ${stats.hasReflection.toString().padStart(3)} / ${total} (${(stats.hasReflection/total*100).toFixed(1)}%)`);
  console.log(`✓ Success metric/measurement:   ${stats.hasMetric.toString().padStart(3)} / ${total} (${(stats.hasMetric/total*100).toFixed(1)}%)`);
  console.log(`✓ Appropriate word count:       ${stats.wordCountOk.toString().padStart(3)} / ${total} (${(stats.wordCountOk/total*100).toFixed(1)}%)`);
  console.log(`✗ Contains vague language:      ${stats.hasVague.toString().padStart(3)} / ${total} (${(stats.hasVague/total*100).toFixed(1)}%)`);
  console.log(`✓ NO ISSUES FOUND:              ${stats.noIssues.toString().padStart(3)} / ${total} (${(stats.noIssues/total*100).toFixed(1)}%)`);
  console.log();
  
  if (allIssues.length > 0) {
    console.log(`CHALLENGES WITH ISSUES (${allIssues.length} lessons need attention)`);
    console.log('-'.repeat(80));
    
    const showCount = Math.min(25, allIssues.length);
    for (let i = 0; i < showCount; i++) {
      const issue = allIssues[i];
      console.log(`\n${i + 1}. ${issue.title}`);
      console.log(`   Word count: ${issue.wordCount}`);
      for (const iss of issue.issues) {
        console.log(`   ⚠ ${iss}`);
      }
      console.log(`   Challenge: ${issue.challengePreview}`);
    }
    
    if (allIssues.length > showCount) {
      console.log(`\n... and ${allIssues.length - showCount} more challenges with issues`);
    }
  } else {
    console.log('✓ ALL CHALLENGES PASS QUALITY CHECKS');
  }
  
  console.log();
  console.log('='.repeat(80));
  console.log('RECOMMENDATIONS');
  console.log('='.repeat(80));
  console.log();
  
  if (stats.hasImmediacy < total * 0.9) {
    console.log('• Add immediacy triggers (right now, today, next time) to more challenges');
  }
  
  if (stats.hasTimebound < total * 0.85) {
    console.log('• Make challenges more time-bound (<10 minutes ideal)');
  }
  
  if (stats.hasReflection < total * 0.85) {
    console.log('• Add reflection prompts (notice what happens, observe, pay attention)');
  }
  
  if (stats.hasVague > total * 0.1) {
    console.log('• Replace vague language (try to, work on) with specific actions');
  }
  
  if (stats.hasMetric < total * 0.5) {
    console.log('• Add clearer success criteria so users know when they\'ve completed the challenge');
  }
  
  if (stats.noIssues >= total * 0.8) {
    console.log('✓ Challenges are in excellent shape - most meet all quality criteria');
  } else if (stats.noIssues >= total * 0.6) {
    console.log('⚠ Most challenges are good, but some need attention');
  } else {
    console.log('⚠ Significant number of challenges need enhancement');
  }
  
  console.log();
  console.log('='.repeat(80));
  console.log('COMPLETION STATUS');
  console.log('='.repeat(80));
  console.log();
  console.log(`Overall Quality Score: ${(stats.noIssues/total*100).toFixed(1)}% of challenges have no issues`);
  console.log();
  
  // Return stats for summary
  return { total, stats, issuesCount: allIssues.length };
}

function main() {
  try {
    const lessons = extractLessonsFromSQL('ascevo/supabase/lessons_all_pillars.sql');
    console.log(`Extracted ${lessons.length} lessons from SQL file\n`);
    
    const results = generateReport(lessons);
    
    // Write detailed results to file
    const reportPath = '.kiro/specs/lesson-content-enhancement/CHALLENGE_REVIEW_REPORT.md';
    const reportContent = `# Challenge Card Actionability Review

**Date**: ${new Date().toISOString().split('T')[0]}
**Total Lessons Analyzed**: ${results.total}

## Summary

- **Challenges with no issues**: ${results.stats.noIssues} / ${results.total} (${(results.stats.noIssues/results.total*100).toFixed(1)}%)
- **Challenges needing attention**: ${results.issuesCount}

## Quality Metrics

| Metric | Count | Percentage |
|--------|-------|------------|
| Immediacy trigger present | ${results.stats.hasImmediacy} | ${(results.stats.hasImmediacy/results.total*100).toFixed(1)}% |
| Time-bound language | ${results.stats.hasTimebound} | ${(results.stats.hasTimebound/results.total*100).toFixed(1)}% |
| Clear action verb | ${results.stats.hasActionVerb} | ${(results.stats.hasActionVerb/results.total*100).toFixed(1)}% |
| Reflection prompt | ${results.stats.hasReflection} | ${(results.stats.hasReflection/results.total*100).toFixed(1)}% |
| Success metric/measurement | ${results.stats.hasMetric} | ${(results.stats.hasMetric/results.total*100).toFixed(1)}% |
| Appropriate word count (30-100) | ${results.stats.wordCountOk} | ${(results.stats.wordCountOk/results.total*100).toFixed(1)}% |

## Status

${results.stats.noIssues >= results.total * 0.8 ? '✅ **PASS** - Challenges are in excellent shape' : 
  results.stats.noIssues >= results.total * 0.6 ? '⚠️ **NEEDS ATTENTION** - Most challenges are good, but some need work' :
  '❌ **NEEDS WORK** - Significant number of challenges need enhancement'}

## Next Steps

${results.issuesCount > 0 ? `Review the ${results.issuesCount} challenges flagged in the analysis output.` : 'All challenges meet quality criteria.'}
`;
    
    fs.writeFileSync(reportPath, reportContent);
    console.log(`\nDetailed report written to: ${reportPath}`);
    
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

main();
